src/test/sql/sql_pipeline_test.cpp

```diff
@@ -0,0 +1,385 @@
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_TRUE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPNotValidated) {
+  SQLPipeline sql_pipeline(_select_query_a, false);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_FALSE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQP) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  sql_pipeline.get_optimized_logical_plan();
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPWithJoinFilter) {
+  SQLPipeline sql_pipeline(_join_query);
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  // TODO(anyone): Find an example where the optimized version is different than the unoptimized one.
+  //  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+  //  EXPECT_EQ(lqp_roots.at(0)->left_child()->type(), LQPNodeType::Predicate);
+  //  EXPECT_EQ(lqp_roots.at(0)->left_child()->left_child()->type(), LQPNodeType::Join);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPInvalid) {
+  SQLPipeline sql_pipeline(_invalid_sql);
+  EXPECT_THROW(sql_pipeline.get_optimized_logical_plan(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPValidated) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_TRUE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPNotValidated) {
+  SQLPipeline sql_pipeline(_select_query_a, false);
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_FALSE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetQueryPlan) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& plan = sql_pipeline.get_query_plan();
+  const auto& roots = plan.tree_roots();
+
+  EXPECT_EQ(roots.size(), 1ul);
+  EXPECT_EQ(roots.at(0)->name(), "Projection");
+}
+
+TEST_F(SQLPipelineTest, GetQueryPlanTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  sql_pipeline.get_query_plan();
+  const auto& plan = sql_pipeline.get_query_plan();
+  const auto& roots = plan.tree_roots();
+
+  EXPECT_EQ(roots.size(), 1ul);
+  EXPECT_EQ(roots.at(0)->name(), "Projection");
+}
+
+TEST_F(SQLPipelineTest, GetQueryPlanJoinWithFilter) {
```

Comment:
dito, please don't rely on Operator structure, this might change in the future.

Reply:
Any idea how to test this without relying on the operators? I think having tests for this is very valuable and just because we would have to change them if we changed the optimizers shouldn't prevent us from them.